src/rebootstrap-raft/main.go

```diff
@@ -0,0 +1,290 @@
+
+	logWriter := &loggoWriter{logger, loggo.DEBUG}
+	raftConfig.Logger = log.New(logWriter, "", 0)
+
+	if err := raft.ValidateConfig(raftConfig); err != nil {
+		return nil, errors.Annotate(err, "validating raft config")
+	}
+	return raftConfig, nil
+}
+
+func makeRaftServers(members []replicaset.Member, apiPort int) (raft.Configuration, error) {
+	var empty raft.Configuration
+	var servers []raft.Server
+	for _, member := range members {
+		id, ok := member.Tags[jujuMachineKey]
+		if !ok {
+			return empty, errors.NotFoundf("juju machine id for replset member %d", member.Id)
+		}
+		baseAddress, _, err := net.SplitHostPort(member.Address)
+		if err != nil {
+			return empty, errors.Annotatef(err, "getting base address for replset member %d", member.Id)
+		}
+		apiAddress := net.JoinHostPort(baseAddress, strconv.Itoa(apiPort))
+		suffrage := raft.Voter
+		if member.Votes != nil && *member.Votes < 1 {
+			suffrage = raft.Nonvoter
+		}
+		server := raft.Server{
+			ID:       raft.ServerID(id),
+			Address:  raft.ServerAddress(apiAddress),
+			Suffrage: suffrage,
+		}
+		servers = append(servers, server)
+	}
+	return raft.Configuration{Servers: servers}, nil
+}
+
+// NewLogStore opens a boltDB logstore in the specified directory. If
+// the directory doesn't already exist it'll be created.
+func NewLogStore(dir string) (*raftboltdb.BoltStore, error) {
+	if err := os.MkdirAll(dir, 0700); err != nil {
+		return nil, errors.Trace(err)
+	}
+	logs, err := raftboltdb.New(raftboltdb.Options{
+		Path: filepath.Join(dir, "logs"),
+	})
+	if err != nil {
+		return nil, errors.Annotate(err, "failed to create bolt store for raft logs")
+	}
+	return logs, nil
+}
+
+// NewSnapshotStore opens a file-based snapshot store in the specified
+// directory. If the directory doesn't exist it'll be created.
+func NewSnapshotStore(
+	dir string,
+	retain int,
+) (raft.SnapshotStore, error) {
+	const logPrefix = "[snapshot] "
+	if err := os.MkdirAll(dir, 0700); err != nil {
+		return nil, errors.Trace(err)
+	}
+	logWriter := &loggoWriter{logger, loggo.DEBUG}
+	logLogger := log.New(logWriter, logPrefix, 0)
+
+	snaps, err := raft.NewFileSnapshotStoreWithLogger(dir, retain, logLogger)
+	if err != nil {
+		return nil, errors.Annotate(err, "failed to create file snapshot store")
+	}
+	return snaps, nil
+}
+
+// loggoWriter is an io.Writer that will call the embedded
+// logger's Log method for each Write, using the specified
+// log level.
+type loggoWriter struct {
+	logger loggo.Logger
+	level  loggo.Level
+}
+
+// Write is part of the io.Writer interface.
+func (w *loggoWriter) Write(p []byte) (int, error) {
+	w.logger.Logf(w.level, "%s", p[:len(p)-1]) // omit trailing newline
+	return len(p), nil
+}
+
+func (c *rebootstrapCommand) dial() (*mgo.Session, error) {
+	info := &mgo.DialInfo{
+		Addrs: []string{net.JoinHostPort(c.hostname, c.mongoPort)},
+	}
+	if c.username != "" {
+		info.Database = "admin"
+		info.Username = c.username
+		info.Password = c.password
```

Comment:
should we warn or error if password is not supplied?

Reply:
Yeah, I think this and machineID should be required parameters.